File: leetcode/maximum-subarray.java
// O(n) time, O(n) space
class Solution {
    public int maxSubArray(int[] nums) {
        int a[] = new int[nums.length];
        a[0] = nums[0];
        int max = a[0];
        
        for(int i = 1; i < nums.length; i++) {
            a[i] = Math.max(nums[i], a[i-1]+nums[i]);
            max = Math.max(a[i], max);
        }
        
        return max;
    }
}

// Solution without storing all the previous max sub-arrays
// * A little harder to understand
// O(n) time (faster than 100%), O(1) space
class Solution {
    public int maxSubArray(int[] nums) {
        if(nums.length == 0)
            return 0;
        if(nums.length == 1)
            return nums[0];

        int runSum = nums[0]; // The maximum sum of the array from 0 to i
        int max = runSum;

        for(int i = 1; i < nums.length; i++) {
            runSum = Math.max(nums[i], runSum+nums[i]);
            max = Math.max(max, runSum);
        }
        
        return max;
    }
}

File: leetcode/maximum-subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        """
        Classic DP problem
        
        Solution 1: BF DFS
        Iterate over all of the possible subarrays and compute their sum, storing
        some max_sum that we return
        time: O(n^2), since we examine all subarrays
        space: O(1)
        
        Solution 2: BF DFS with caching
        Go from left to right
            
        
        """
        dp = nums[:]
        dp[0] = nums[0]
        
        for i in range(1, len(nums)):
            dp[i] = max(dp[i], dp[i]+dp[i-1])        
        
        return max(dp)
